res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+        constructor(options) {
+            options.outKey = "sync_enabled";
+            super(options);
+        }
+        unshift() {
+            this.input = function(channel, control, value, status, _group) {
+                if (this.isPress(channel, control, value, status)) {
+                    if (engine.getValue(this.group, "sync_enabled") === 0) {
+                        engine.setValue(this.group, "beatsync", 1);
+                        this.longPressTimer = engine.beginTimer(this.longPressTimeout, () => {
+                            engine.setValue(this.group, "sync_enabled", 1);
+                            this.longPressTimer = 0;
+                        }, true);
+                    } else {
+                        engine.setValue(this.group, "sync_enabled", 0);
+                    }
+                } else {
+                    if (this.longPressTimer !== 0) {
+                        engine.stopTimer(this.longPressTimer);
+                        this.longPressTimer = 0;
+                    }
+                }
+            };
+        }
+        shift() {
+            this.inKey = "quantize";
+            this.type = Button.types.toggle;
+            this.input = Button.prototype.input;
+        }
+    };
+
+    const LoopToggleButton = class extends Button {
+        constructor(options) {
+            options.inKey = "reloop_exit";
+            options.outKey = "loop_enabled";
+            super(options);
+        }
+        inValueScale() {
+            return 1;
+        }
+    };
+
+    const HotcueButton = class extends Button {
+        constructor(options) {
+            if (options.number === undefined) {
+                throw "No hotcue number specified for HotcueButton constructor";
+            }
+            if (options.colorMapper !== undefined || options.sendRGB !== undefined) {
+                options.colorKey = "hotcue_" + options.number + "_color";
+            }
+            options.outKey = "hotcue_" + options.number + "_enabled";
+            super(options);
+        }
+        unshift() {
+            this.inKey = "hotcue_" + this.number + "_activate";
+        }
+        shift() {
+            this.inKey = "hotcue_" + this.number + "_clear";
+        }
+        output(value) {
+            const outval = this.outValueScale(value);
+            // NOTE: outputColor only handles hotcueColors
+            // and there is no hotcueColor for turning the LED
+            // off. So the `send()` function is responsible for turning the
+            // actual LED off.
+            if (this.colorKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorKey));
+            } else {
+                this.send(outval);
+            }
+        }
+        outputColor(colorCode) {
+            // Sends the color from the colorCode to the controller. This
+            // method will not be called if no colorKey has been specified.
```

Comment:
This comment is true now, but in general it's unsafe to make assumptions about how a function will be called if the prerequisites aren't enforced.  What if someone's controller script calls this function without a colorKey defined? If that's a problem, then we should check for colorKey != undefined and return if so.

Reply:
Well this is assumed here because iirc, this method is essentially private. No user should call this themselves because the colorCode is an opaque type. I don't want to introduce much checks because this method is much hotter than most other parts of a component.

Reply:
Maybe prepend the function name with _ to indicate that it's "private"? That seems to be the convention in js-land.

Reply:
ok, no need to add the undefined check

Reply:
Sure, actually JS has support for members, but its too new of an ES version so we'll have to wait for [QTBUG-97724](https://bugreports.qt.io/browse/QTBUG-97724)

Reply:
yeah underscore is not enforced but at least it's a signal